core/drivers/imx/dcp/dcp.c

```diff
@@ -0,0 +1,773 @@
+	struct dcp_align_buf input = { };
+	uint32_t offset = 0;
+	uint32_t nb_blocks = 0;
+	size_t size_todo = 0;
+	size_t size_left = 0;
+
+	if (!hashdata || !data || !len)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	if (!hashdata->initialized) {
+		EMSG("hashdata is uninitialized");
+		return TEE_ERROR_BAD_STATE;
+	}
+
+	/* Get number of blocks */
+	nb_blocks = (hashdata->ctx_size + len) / DCP_SHA_BLOCK_SIZE;
+	size_todo = nb_blocks * DCP_SHA_BLOCK_SIZE;
+	size_left = len - size_todo + hashdata->ctx_size;
+	desc = &hashdata->dcp_data.desc;
+
+	if (size_todo) {
+		/* Allocate buffer as input */
+		ret = dcp_calloc_align_buf(&input, size_todo);
+		if (ret != TEE_SUCCESS)
+			return TEE_ERROR_OUT_OF_MEMORY;
+
+		/* Copy previous data if any */
+		offset = size_todo - hashdata->ctx_size;
+		memcpy(input.data, hashdata->ctx.data, hashdata->ctx_size);
+		memcpy(input.data + hashdata->ctx_size, data, offset);
+		hashdata->ctx_size = 0;
+
+		desc->src_buffer = (uint8_t *)input.paddr;
+		desc->buff_size = size_todo;
+
+		cache_operation(TEE_CACHECLEAN, input.data, size_todo);
+
+		ret = dcp_run(&hashdata->dcp_data);
+		desc->ctrl0 &= ~DCP_CONTROL0_HASH_INIT;
+
+		dcp_free(&input);
+	} else {
+		size_left = len;
+		offset = 0;
+		ret = TEE_SUCCESS;
+	}
+
+	/* Save any data left */
+	memcpy(hashdata->ctx.data + hashdata->ctx_size, data + offset,
+	       size_left);
+	hashdata->ctx_size += size_left;
+
+	return ret;
+}
+
+TEE_Result dcp_sha_do_final(struct dcp_hash_data *hashdata, uint8_t *digest,
+			    size_t digest_size)
+{
+	TEE_Result ret = TEE_ERROR_GENERIC;
+	size_t payload_size = 0;
+	struct dcp_descriptor *desc = NULL;
+	struct dcp_align_buf payload = { };
+
+	if (!hashdata || !digest)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	if (!hashdata->initialized) {
+		EMSG("hashdata is uninitialized");
+		return TEE_ERROR_BAD_STATE;
+	}
+
+	if (digest_size < hash_alg[hashdata->alg].size) {
+		EMSG("Digest buffer size is to small, should be %" PRId32,
+		     hash_alg[hashdata->alg].size);
+		return TEE_ERROR_BAD_PARAMETERS;
+	}
+
+	desc = &hashdata->dcp_data.desc;
+	payload_size = hash_alg[hashdata->alg].size;
+
+	/* Handle the case where the input message is NULL */
+	if ((desc->ctrl0 & DCP_CONTROL0_HASH_INIT) && hashdata->ctx_size == 0) {
+		if (hashdata->alg == DCP_SHA1)
+			memcpy(digest, sha1_null_msg, payload_size);
+		if (hashdata->alg == DCP_SHA256)
+			memcpy(digest, sha256_null_msg, payload_size);
+		ret = TEE_SUCCESS;
+	} else {
+		/* Allocate buffer for the digest */
+		ret = dcp_calloc_align_buf(&payload, payload_size);
+		if (ret != TEE_SUCCESS)
+			return TEE_ERROR_OUT_OF_MEMORY;
+
+		/* Set work packet for last iteration */
+		desc->ctrl0 |= DCP_CONTROL0_HASH_TERM;
+		desc->src_buffer = (uint8_t *)hashdata->ctx.paddr;
+		desc->buff_size = hashdata->ctx_size;
+		desc->payload = (uint8_t *)payload.paddr;
+
+		cache_operation(TEE_CACHECLEAN, hashdata->ctx.data,
+				hashdata->ctx_size);
+		cache_operation(TEE_CACHEFLUSH, payload.data, payload_size);
```

Comment:
Why `TEE_CACHEFLUSH` instead of `TEE_CACHEINVALIDATE`?

Reply:
A ```TEE_CACHEINVALIDATE``` is enough